### PR TITLE
Feature table clustering

### PR DESCRIPTION
**Pull request by cceasy**
## Summary
This PR adds support for using clustering columns instead of partition columns when querying Delta Lake tables. This enables efficient data pruning on tables that use Delta Lake's clustering feature rather than traditional Hive-style partitioning.

## Why
Delta Lake tables can be clustered by timestamp columns (e.g., created_at) instead of using Hive-style partition columns (e.g., ds). When querying these tables, filtering by a derived partition column causes full table scans because Delta Lake cannot leverage its clustering optimization.

## Usage Example
```
source = Source(
    events=EventSource(
        table="db.clustered_events",
        query=Query(
            selects=selects("user_id", "event_type", "value"),
            time_column="unix_millis(CAST(created_at AS TIMESTAMP))",
            clustering_column="created_at",
            clustering_format="yyyy-MM-dd",
        ),
    ),
)
```
This generates WHERE clauses like:
  - created_at >= '2024-01-01'
  - created_at < '2024-01-02'

Instead of the previous behavior which would try to filter on a non-existent ds column.

## Checklist
- [ ] Added Unit Tests
- [ ] Covered by existing CI
- [ ] Integration tested
- [ ] Documentation update

## Summary by CodeRabbit

* **New Features**
  * Added optional clustered flag for queries to support clustered (e.g., Delta Lake) tables.
  * Partition filtering now uses clustering-aware predicates to ensure correct scanning of clustered data.

* **Tests**
  * Added tests covering clustered and non-clustered query behavior and partition predicate edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by nikhil-zlai — CHANGES_REQUESTED** [review]
Thanks for making the change. I think we could simplify both the API and the change further.

I think the change would be a lot simpler if you added a boolean flag "clustered=true"
instead of adding "cluster_format" and "cluster_column" 

if clustered=true, we will re-use the partition_format and partition_column - but simply generate different filter expression based on the clustered flag

remember to NOT set the default thrift value for the flag.

**Comment by cceasy** [comment]
> Thanks for making the change. I think we could simplify both the API and the change further.
> 
> I think the change would be a lot simpler if you added a boolean flag "clustered=true" instead of adding "cluster_format" and "cluster_column"
> 
> if clustered=true, we will re-use the partition_format and partition_column - but simply generate different filter expression based on the clustered flag
> 
> remember to NOT set the default thrift value for the flag.

Sure, updated

File: spark/src/main/scala/ai/chronon/spark/catalog/TableUtils.scala [hunk 1]
```diff
@@ -648,10 +658,40 @@ class TableUtils(@transient val sparkSession: SparkSession) extends Serializable
...
+    *
+    * For partition columns (default): uses inclusive end (column <= 'value')
+    * For clustering columns: uses exclusive end with next interval (column < 'next(value)')
+    *   This ensures proper timestamp range filtering for Delta Lake clustering.
+    *
+    * @param range The partition range with start/end dates
+    * @param partitionColumn Optional partition column name. Defaults to class-level partitionColumn if None.
+    * @param clusteringColumn Optional clustering column name. If provided, uses clustering semantics.
+    * @return Sequence of WHERE clause strings
+    */
```
[the remark this answers is not in the document]
Comment (cceasy) [reply spark/src/main/scala/ai/chronon/spark/catalog/TableUtils.scala]: Yeah, I can remove it since this function code is clear enough to read and function body also has a brief comment.

File: spark/src/main/scala/ai/chronon/spark/catalog/TableUtils.scala [hunk 2]
```diff
@@ -648,10 +648,24 @@ class TableUtils(@transient val sparkSession: SparkSession) extends Serializable
...
+      range: PartitionRange,
+      partitionColumn: Option[String] = None,
+      clustered: Boolean = false
+  ): Seq[String] = {
+    val col = partitionColumn.getOrElse(this.partitionColumn)
+    if (clustered) {
+      // For clustered tables (timestamp/timestamp_ntz), use >= start and < next(end)
+      // This ensures we include all data up to but not including the next interval
+      val startClause = Option(range.start).map(s => s"$col >= '$s'")
+      val endClause = Option(range.end).map(e => s"$col < '${range.partitionSpec.after(e)}'")
```
[the remark this answers is not in the document]
Comment (nikhil-zlai) [reply spark/src/main/scala/ai/chronon/spark/catalog/TableUtils.scala]: true!

File: spark/src/main/scala/ai/chronon/spark/catalog/TableUtils.scala [hunk 3]
```diff
@@ -648,10 +648,24 @@ class TableUtils(@transient val sparkSession: SparkSession) extends Serializable
...
-    val startClause = Option(partitionRange.start).map(s"$partitionColumn >= '" + _ + "'")
-    val endClause = Option(partitionRange.end).map(s"$partitionColumn <= '" + _ + "'")
-    (startClause ++ endClause).toSeq
+  def whereClauses(
+      range: PartitionRange,
+      partitionColumn: Option[String] = None,
+      clustered: Boolean = false
+  ): Seq[String] = {
+    val col = partitionColumn.getOrElse(this.partitionColumn)
+    if (clustered) {
```
Comment (nikhil-zlai) [inline spark/src/main/scala/ai/chronon/spark/catalog/TableUtils.scala]: we can keep the clustered arg around - but we should still not have a branch here I think.

**Review by nikhil-zlai — APPROVED** [review]
Thanks for working on this!

**Comment by cceasy** [comment]
@nikhil-zlai can we merge this MR so that I could rebase the latest main branch on my local
